First rendering

anne-gropler opened the pull request:
This pull request implements some features from Issue #36, namely the interactive colouring (font color, background color).
Starting the application llassetgen-rendering, you shoud see something like this (different colours).
![pr](https://user-images.githubusercontent.com/5690856/36484451-a457f06c-1719-11e8-9707-ee895fb97f3b.png)


Why did I separately write the classes `Window `and `WindowQt`, instead of merging them into one?
That's because glbinding is not compatible with gl.h, which might get included by some Qt-classes. Merging `Window `and `WindowQt` throws the error "glbinding is not compatible with gl.h" located in glbinding/nogl.h. To fix this, the GL-rendering we separate the gl from the Qt using those 2 classes, which is encouraged by glbinding authors.

HELP WANTED: Please tell me if the shaders and the fontatlas are loaded from the correct path on other OS (file structure might be different).

WIP: fix CI dependencies.

anne-gropler committed:
created app that uses Qt to display image

anne-gropler committed:
removed debug code

anne-gropler committed:
make appveyor build using Qt

anne-gropler committed:
fix typo: missing )

anne-gropler committed:
Update travis: qt libs

anne-gropler committed:
Update CMakeList to use Qt Version 5.2

Linux qt5-default is 5.2

anne-gropler committed:
APPVEYOR_BUILD_WORKER_IMAGE instead of image

Felerius committed:
Configure cmake to find qt on macOS

anne-gropler committed:
travis: qt5 instead of qt

anne-gropler committed:
travis: don't override but append cmake prefix path

anne-gropler committed:
travis: write qt ref in PATH

anne-gropler committed:
travis: manipulate PATH in global env

anne-gropler committed:
qt + openGl from globjects::qtexample

anne-gropler committed:
removed unnecessary cpplocate

anne-gropler committed:
removed datapath.inl which uses cpplocate

anne-gropler committed:
use glyph texture in shader for glyph rendering (simplified)

anne-gropler committed:
Merge branch 'master' into first-rendering

anne-gropler committed:
update travis dependencies

anne-gropler committed:
added -y option for linux cginternals package

anne-gropler committed:
cppcheck, clang-tidy, check, format

anne-gropler committed:
first attempt in GUI input to render color (still layout problem)

anne-gropler committed:
GUI looks as expected

anne-gropler committed:
user can update background color using qt gui

anne-gropler committed:
restructure: put GUI stuff into another function

anne-gropler committed:
user can use gui to change font color

anne-gropler committed:
non-functional changes due to cppcheck, clang etc.

anne-gropler committed:
added testpicture

anne-gropler committed:
consistent using of path for loading data

anne-gropler committed:
non-functional changes: documentation

anne-gropler committed:
fixed typo

anne-gropler committed:
non-functional changes: class doc

anne-gropler committed:
CI: travis

anne-gropler committed:
typo fix

anne-gropler committed:
travis macOS: add glbinding globjects to cmake prefix path

anne-gropler committed:
new version of globjects (2.0) needs other find scripts for glm

anne-gropler committed:
ported code using globjects from globjects-v1 to globjects-v2

anne-gropler committed:
format: removed extra ';'

anne-gropler committed:
travis osx: glbinding included in globjects?

anne-gropler committed:
revert last changes; typo fix

anne-gropler committed:
travis: add glbinding globjects location to PATH

anne-gropler committed:
namespace fix

anne-gropler committed:
travis osx: add cginternals tap

anne-gropler committed:
travis osx: automatically export install locations of glbinding and globjects to PATH

anne-gropler committed:
added override keyword to overriding functions

anne-gropler committed:
travis: debug echoes

anne-gropler committed:
travis read different command.. why? changing : to ,

anne-gropler committed:
seems that there is no link from usr/local/opt to usr/local/Cellar ?

anne-gropler committed:
PATH is not exportet in bevore_install, so I set it manually.

anne-gropler committed:
fix: mark overriding function with override.

anne-gropler commented:
@scheibel 
Travis can't build on Mac OS because homebrew installs globjects v1 instead of v2.
apt get on linux installs v2.
Appveyour can't build on windows because there is no NuGet nor Chocolatey for glbinding and globjects.

Independently from that, I'd appreciate reviews for this PR.

Felerius commented on .travis.yml:
```diff
@@ -14,6 +14,10 @@ compiler:
     - CMAKE_OPTIONS="-DOPTION_BUILD_EXAMPLES=On"
+    - PATH="/usr/local/opt/qt/bin:$PATH"
+    - PATH="/usr/local/opt/glbinding:$PATH"
+    - PATH="/usr/local/opt/globjects:$PATH"
```
I would put these all into one line, otherwise the list format suggests that three separate variables are being set

anne-gropler replied:
I'd do that, too.. actually I was confused about the ":", I expected a ";", so I just copied the line to make sure I won't make syntax mistakes :D I'm not used to linux nor macOS

Felerius commented on .travis.yml:
```diff
@@ -24,9 +28,13 @@ matrix:
   # libpng is already installed by default
-  - if [ $TRAVIS_OS_NAME == osx ]; then brew update && brew install cppcheck freetype; fi
+  - if [ $TRAVIS_OS_NAME == osx ]; then brew tap cginternals/homebrew-cginternals; fi
+  - if [ $TRAVIS_OS_NAME == osx ]; then brew update && brew install cppcheck freetype qt5 glm glbinding globjects; export CMAKE_PREFIX_PATH="$CMAKE_PREFIX_PATH;$(brew --prefix qt);$(brew --prefix glbinding);$(brew --prefix globjects)"; fi
```
I think this could be turned into a multi line string like so (to avoid repeating the if statement making the whole thing more readable):
```yaml
- |-
  if [ ... ]; then
    ...;
    ...;
  fi
```

The `|` starts a multi line string, and the additional `-` cuts out trailing newlines from the string (not necessarily needed).

anne-gropler replied:
Cool, thanks!

Felerius commented on .travis.yml:
```diff
@@ -24,9 +28,13 @@ matrix:
+  - if [ $TRAVIS_OS_NAME == linux ]; then sudo apt-add-repository ppa:cginternals/ppa -y; fi
+  - if [ $TRAVIS_OS_NAME == linux ]; then sudo apt-add-repository ppa:cginternals/backports-ppa -y; fi
+  - if [ $TRAVIS_OS_NAME == linux ]; then sudo apt-get update -qq; fi
+  - if [ $TRAVIS_OS_NAME == linux ]; then sudo apt-get install -qq cmake3 cppcheck clang-tidy-3.8 libfreetype6-dev libpng-dev qt5-default libglm-dev libglbinding-dev libglobjects-dev; fi
```
As with below, the repeated if statements make this kind of hard to read, but we can update this to use the alternative way of specifying apt dependencies once #39 with 7d13192 gets merged.

Felerius commented on appveyor.yml:
```diff
@@ -45,13 +45,23 @@ install:
+- if "%APPVEYOR_BUILD_WORKER_IMAGE%"=="Visual Studio 2017" ( set Qt5Core_DIR=C:\Qt\5.9.2\msvc2017_64\lib\cmake\Qt5Core)
+- if "%APPVEYOR_BUILD_WORKER_IMAGE%"=="Visual Studio 2017" ( set Qt5Gui_DIR=C:\Qt\5.9.2\msvc2017_64\lib\cmake\Qt5Gui)
+- if "%APPVEYOR_BUILD_WORKER_IMAGE%"=="Visual Studio 2017" ( set Qt5OpenGL_DIR=C:\Qt\5.9.2\msvc2017_64\lib\cmake\Qt5OpenGL)
+- if "%APPVEYOR_BUILD_WORKER_IMAGE%"=="Visual Studio 2017" ( set Qt5Widgets_DIR=C:\Qt\5.9.2\msvc2017_64\lib\cmake\Qt5Widgets)
```
I don't know the appveyor yaml syntax very well, but if there is a way do to multiline if statements that would make this way more readable.

anne-gropler replied:
I tried various multi-lines and none of them work. Sorry :(

Felerius commented on source/llassetgen-rendering/CMakeLists.txt:
```diff
@@ -0,0 +1,154 @@
+set(sources
+    main.cpp
+	
+	WindowQt.cpp
```
There is a lot of mixing of tabs and spaces in this file. I think we use spaces in every other `CMakeLists.txt` and should probably spay consistent.

anne-gropler replied:
Thanks. I didn't see that my Editor used tabs.

Felerius commented on source/llassetgen-rendering/main.cpp:
```diff
@@ -0,0 +1,383 @@
+
+#include "WindowQt.h"
+
+using namespace gl;
```
I don't know how many types in this file are from the `gl` namespace, but if its not too many a few `using gl::typename` lines would do the same and not pull in all members of the namespace.

anne-gropler replied:
There are currently 20 identifiers that need that namespace, and they all start with gl. So instead of `GL_TEXTURE_2D` I'd need to write `gl::GL_TEXTURE_2D`; or `glClear becomes `gl::glClear`.
So I'd rather use this namespace as it is. 

Well, I am not sure... are 20 identifiers few or many?

Felerius replied:
Oh, ok. If that namespace just contains the usual opengl symbols (which are prefixed with `GL_`, `gl`, etc. anyway), then the `using namespace` is fine.

scheibel replied:
Yes, we designed the `gl*` namespaces to include only the official symbols from the OpenGL API of their respective version (namespace `gl` includes all).

Felerius commented on source/llassetgen-rendering/main.cpp:
```diff
@@ -0,0 +1,383 @@
+        m_samplerIndex = 0;
+        m_texture = nullptr;
+        m_backgroundColor = glm::vec4(1.f, 1.f, 1.f, 1.f);
+        m_fontColor = glm::vec4(0.f, 0.f, 0.f, 1.f);
```
As far as i know it is preferred to initialize these directly where the member variable is declared (i.e. `glm::vec4 m_fontColor {0.f, 0.f, 0.f, 1.f}` at the end of the class)

anne-gropler replied:
Ok :)

Felerius commented on source/llassetgen-rendering/main.cpp:
```diff
@@ -0,0 +1,383 @@
+
+    glm::vec4 m_backgroundColor;
+    glm::vec4 m_fontColor;
+    int m_samplerIndex;
```
Is the `m_` prefix there to stay consistent with Qt or any other library? Otherwise I would use the same style we use everywhere else.

anne-gropler replied:
m_ stands for member variable, it's a coding style we use for years at the CGS, that's why I automatically used it, sorry :D
So, as far as I see, the style in this project is to simply remove the m_?

Felerius commented on source/llassetgen-rendering/main.cpp:
```diff
@@ -0,0 +1,383 @@
+        // TODO load using own png loader instead of Qt (blocked: wait for this feature to be merged into master, then
+        // pull). TODO: Make sure this file structure is the same on other OS
+        auto path = QApplication::applicationDirPath();
+        auto *image = new QImage(path + "/../../data/llassetgen-rendering/testfontatlas_rgb.png");
```
If we want to include a default font atlas long-term, we should probably include it in the executable or figure out how to install into and reference the standard directories (stuff like this would probably go into `/usr/[local/]share` on Linux and macOS).

anne-gropler replied:
This path finds the executable path and works relative from there.
It's not about including a default atlas long-term, but asking the user to specify a path where to load from. Until then, we just load a png as a mock. That's the plan :D

Felerius commented on source/llassetgen-rendering/main.cpp:
```diff
@@ -0,0 +1,383 @@
+        m_fragmentShaderSource.reset(nullptr);
+        m_fragmentShaderTemplate.reset(nullptr);
+        m_fragmentShader.reset(nullptr);
+        m_vao.reset(nullptr);
```
`variable = nullptr` does the same thing and is a bit more clear in my opinion.

anne-gropler replied:
Since those variables are unique pointers, I'd stick to reset ;)

Felerius commented on source/llassetgen-rendering/main.cpp:
```diff
@@ -0,0 +1,383 @@
+        int blue = value.toInt();
+        m_fontColor.b = blue / 255.f;
+        paint();
+    }
```
Is there a way to deduplicate these methods? As it stands, it is basically the same code copy pasted 6 times. Maybe something like:
```cpp
void applyColorChange(float& color, QString value) {
    color = value.toInt() / 255.f;
    paint();
}

virtual void fontColorBChanged(QString value) override {
    applyColorChange(m_fontColor.b, value);
}
```

anne-gropler replied:
yep, good idea, thanks!

Felerius commented on source/llassetgen-rendering/main.cpp:
```diff
@@ -0,0 +1,383 @@
+    labelB->setMaximumWidth(90);
+    backgroundB->setValidator(colorValidator);
+    backgroundB->setPlaceholderText("255");
+    backgroundB->setMaximumWidth(45);
```
Could probably made shorter by making a factory methods for color fields

anne-gropler replied:
A factory (as in GoF) needs subclasses, I think that's too much here. But I created a method that replaces 3 lines and makes maintenance easier.

Felerius reviewed (approved):
I left a few comments (mostly nitpicky code style related stuff, feel free to ignore some things if you think they are unimportant), but the code generally looks good. I didn't get to test the code because I don't have time to build globjects myself right now (I'm using macOS)

anne-gropler committed:
travis, appveyor: use multiline if-statements

anne-gropler committed:
coding style

anne-gropler committed:
appveyor: multiline seems to work differently...?

anne-gropler committed:
deduplicated color changed functions as Felerius suggested.

anne-gropler committed:
refactoring for color inputs

anne-gropler committed:
travis: use cginternals tap for globjects to get newer version

anne-gropler committed:
automatic format

anne-gropler committed:
appveyor: glbinding globjects via vcpkg

anne-gropler committed:
appveyor: glm

anne-gropler committed:
appveyor has not the latest git

anne-gropler committed:
appveyor: cmake in wrong dir

anne-gropler committed:
appveyor: build cache

anne-gropler committed:
appveyor: multi-line if-statement doesn't work

anne-gropler committed:
debugging appveyor

anne-gropler committed:
debugging appveyor: remove toolchain file param

anne-gropler committed:
debugging appveyor: provide toolchain param AND qt params

anne-gropler committed:
debugging appveyor: change order of cmake params

anne-gropler committed:
Merge branch 'master' into first-rendering

anne-gropler committed:
debugging appveyor: see if Qt5Core_DIR is still set in before_build

anne-gropler committed:
debugging appveyor: different find for Qt

anne-gropler committed:
debugging appveyor: qt5 instead of qt

anne-gropler commented:
@scheibel @cgcostume You can start reviewing. The WIP only applies for the Appveyor dependencies.
Observe https://help.appveyor.com/discussions/problems/12635-using-vcpkg-for-glm-etc-cmake-doesnt-find-the-preinstalled-qt-anymore

anne-gropler committed:
debugging appveyor: see if qt5 is found when no vcpkg is used

anne-gropler committed:
debugging appveyor: qt as cmake param -D

anne-gropler committed:
debugging appveyor: change param order

anne-gropler committed:
prepared Qt event handler

anne-gropler committed:
added transform, still not used

anne-gropler committed:
simplified label creation (by not creating a label manually)

anne-gropler committed:
Merge branch 'first-rendering' into rendering-transforms

Blaidd-Drwg commented:
Re: help wanted: I figure that also depends on where your IDE places the executable. Clion keeps it in `openll-asset-generator/cmake-build-debug`, so I have to delete one `../` from every path to make it work.

anne-gropler commented:
Thanks, that's good to know! I'll try to find a better way for loading files, but that wil be another PR. Hopefully, I won't need it, since loading from files should be done by our image class.

anne-gropler committed:
format

anne-gropler committed:
Merge branch 'master' into first-rendering

anne-gropler committed:
Merge branch 'first-rendering' into rendering-transforms

anne-gropler committed:
implemented rudimentary zooming and panning

anne-gropler committed:
implemented rudimentary rotation

anne-gropler committed:
Merge branch 'master' into first-rendering

# Resolved Conflicts:
#	.travis.yml
#	appveyor.yml

anne-gropler committed:
merge fix: glm is not found

anne-gropler committed:
merge fix: qt5
The imported target "Qt5::OpenGL" references the file "/usr/lib/x86_64-linux-gnu/libQt5OpenGL.so.5.2.1" but this file does not exist. Possible reasons include: * The file was deleted, renamed, or moved to another location. * An install or uninstall procedure did not complete successfully. * The installation package was faulty and contained "/usr/lib/x86_64-linux-gnu/cmake/Qt5OpenGL/Qt5OpenGLConfig.cmake" but not all the files it references.

anne-gropler committed:
Merge branch 'first-rendering' into rendering-transforms

anne-gropler committed:
nearer near plane; allows more zoom

anne-gropler commented:
Changed near plane from 0.1 to 0.0001 as @Blaidd-Drwg requested in an extern conversation.

anne-gropler committed:
Merge branch 'master' into first-rendering

anne-gropler committed:
Merge branch 'first-rendering' into rendering-transforms

anne-gropler committed:
de-stretched the projection and the textured quad

anne-gropler committed:
The ways of the Format are mysterious.

anne-gropler committed:
doc-comment in shader

anne-gropler committed:
superSampling as in OpenLL

anne-gropler committed:
format

anne-gropler committed:
added a distance field png so that others can start the application with the same data.

anne-gropler committed:
tabs to spaces

anne-gropler committed:
newlines at end of file

anne-gropler committed:
switch between rendering distance field and rendering glyphs

anne-gropler committed:
put 3 lines into one for better readability, as suggested by @Felerius

anne-gropler committed:
non-functional change: initializer list instead of glm::vec2 constructor

anne-gropler committed:
removed unnecessary calls to make gl context current

anne-gropler committed:
tabs to spaces

anne-gropler committed:
overriding methods marked with override.

anne-gropler committed:
Merge branch 'first-rendering' into rendering-transforms

anne-gropler committed:
message when image is not loaded successfully.

anne-gropler committed:
removed unused files

anne-gropler committed:
Merge branch 'rendering-transforms' into rendering-supersampling

# Resolved Conflicts:
#	data/llassetgen-rendering/shader.frag
#	data/llassetgen-rendering/shader.vert

anne-gropler committed:
Merge branch 'rendering-supersampling' into rendering-switchDTs

# Resolved Conflicts:
#	data/llassetgen-rendering/shader.frag

anne-gropler committed:
Merge pull request #44 from hpicgs/rendering-transforms

Rendering transforms

anne-gropler committed:
Merge branch 'master' into first-rendering

anne-gropler committed:
commented lines that are deprecated and throw errors.

anne-gropler committed:
Merge branch 'first-rendering' into rendering-supersampling

anne-gropler committed:
Merge branch 'rendering-supersampling' into rendering-switchDTs

anne-gropler committed:
using C++ unsigned int instead of Qt uint

anne-gropler committed:
Merge branch 'rendering-supersampling' into rendering-switchDTs

anne-gropler committed:
Merge pull request #48 from hpicgs/rendering-supersampling

Rendering supersampling

anne-gropler committed:
added input for DT threshold

anne-gropler committed:
shader fix from last cherry-pick

anne-gropler committed:
Merge branch 'master' into first-rendering

anne-gropler committed:
Merge branch 'first-rendering' of https://github.com/hpicgs/openll-asset-generator into first-rendering

anne-gropler committed:
Merge branch 'first-rendering' into rendering-differentDTs

anne-gropler committed:
comment explaining bad navigation

anne-gropler committed:
prepared GUI elements for DT options

anne-gropler committed:
Merge pull request #49 from hpicgs/rendering-switchDTs

switch between rendering distance field and rendering glyphs

anne-gropler committed:
Merge branch 'master' into first-rendering

anne-gropler committed:
Merge branch 'first-rendering' of https://github.com/hpicgs/openll-asset-generator into first-rendering

anne-gropler committed:
Merge branch 'first-rendering' into rendering-differentDTs

anne-gropler committed:
fixed GUI for dynamic range

anne-gropler committed:
more prepared GUI for Distance Field Creation

anne-gropler committed:
refactored color inputs

anne-gropler committed:
membervariables for distance transform parameters; export button

anne-gropler committed:
appveyor: change param order

anne-gropler committed:
Merge branch 'atlas-creation' into first-rendering

anne-gropler committed:
Merge branch 'atlas-creation' into rendering-differentDTs

anne-gropler committed:
Merge branch 'master' into first-rendering

anne-gropler committed:
Merge branch 'first-rendering' into rendering-differentDTs

anne-gropler committed:
Merge branch 'master' into first-rendering

anne-gropler committed:
Merge branch 'first-rendering' into rendering-differentDTs

anne-gropler committed:
WIP: FontFinder part of llassetgen; trigger altas creation from rendering app

anne-gropler committed:
forgot to include new files, sorry

anne-gropler committed:
reload distance field

anne-gropler committed:
choose: no preset but packing algorithm

Felerius committed:
Assorted fixes for rendering

Image copy/move constructor/assignment
Fontconfig on macOS
...

anne-gropler committed:
some GUI stuff

Felerius committed:
Install fontconfig on travis on macOS

anne-gropler committed:
GUI for padding

anne-gropler committed:
Merge branch 'rendering-differentDTs' of https://github.com/hpicgs/openll-asset-generator into rendering-differentDTs

anne-gropler committed:
updated FontFinde property as in cli branch

anne-gropler committed:
fix from David and Lukas and Justin and Alex and God.

anne-gropler committed:
cpack files for windows installer; still doesn't copy the dlls

Felerius committed:
Don't build CLI as app bundle

anne-gropler committed:
seem to work now

anne-gropler committed:
Merge branch 'master' into windows-installer

# Conflicts:
#	AUTHORS

anne-gropler committed:
habemus logos

anne-gropler committed:
new line at end of file

anne-gropler committed:
Merge branch 'master' into first-rendering

# Conflicts:
#	.travis.yml

anne-gropler committed:
Merge branch 'first-rendering' into rendering-differentDTs

# Conflicts:
#	.travis.yml
#	source/llassetgen/CMakeLists.txt
#	source/llassetgen/include/llassetgen/Atlas.h
#	source/llassetgen/include/llassetgen/FontFinder.h
#	source/llassetgen/include/llassetgen/llassetgen.h
#	source/llassetgen/source/FontFinder.cpp
#	source/llassetgen/source/Image.cpp

anne-gropler committed:
wrong new line

anne-gropler committed:
Merge branch 'first-rendering' into rendering-differentDTs

# Conflicts:
#	.travis.yml

anne-gropler committed:
fix merge error

anne-gropler committed:
fix merge error 2

anne-gropler committed:
Merge branch 'master' into first-rendering

anne-gropler committed:
Merge branch 'first-rendering' into rendering-differentDTs

# Conflicts:
#	source/llassetgen/include/llassetgen/Atlas.h

anne-gropler committed:
fix downsampling

anne-gropler committed:
export FNT, fix minor stuff

anne-gropler committed:
format

anne-gropler committed:
Merge branch 'master' into windows-installer

anne-gropler committed:
Merge branch 'windows-installer' into rendering-differentDTs

anne-gropler committed:
fixed data path

anne-gropler committed:
version number 1.0.0

anne-gropler committed:
removed duplicated dataPath

anne-gropler committed:
Merge pull request #63 from hpicgs/homebrew-formula

Necessary changes for homebrew formula

anne-gropler committed:
path fix (QString --> std::string)

anne-gropler committed:
cleaner path management (data-dir for shader, appDataLocation for output files)

anne-gropler committed:
version number 1.0.1

anne-gropler committed:
path fix for older qt versions (e.g. for travis ubuntu)

anne-gropler committed:
fixed missing newlines from last merge

anne-gropler committed:
Merge pull request #54 from hpicgs/rendering-differentDTs

Rendering different Distance Transforms

anne-gropler committed:
Merge branch 'develop' into first-rendering

# Conflicts:
#	source/llassetgen/include/llassetgen/Atlas.h
